readbuffer.go

```diff
@@ -100,13 +106,15 @@ func (r *readBuffer) Read(b []byte) (int, error) {
 		now := time.Now()
 		if !r.deadline.IsZero() {
 			if now.After(r.deadline) {
-				return 0, errors.New("deadline exceeded")
+				return 0, ErrReadDeadlineExceeded
 			}
 		}
 
 		var t *time.Timer
 		if !r.deadline.IsZero() {
 			t = time.AfterFunc(r.deadline.Sub(now), func() { r.cond.Broadcast() })
```

Comment:
I know you didn't change this but we should really be locking before calling `r.cond.Broadcast()` otherwise there's a race where the broadcast _could_ be called first, and `r.cond.Wait()` would miss it, potentially leading to a blocked goroutine.

Reply:
> I know you didn't change this but we should really be locking before calling r.cond.Broadcast() otherwise there's a race where the broadcast could be called first, and r.cond.Wait() would miss it, potentially leading to a blocked goroutine.

I agree with that. Should I add that to this patch, or will you create a follow-up PR?

> I couldn't find any use of SetDeadline so it's probably safe to say we're not going to see the issue wrt. deadline in rancher/rancher (for now)

`SetDeadline`/`SetReadDeadline` are in the general `net.Conn` interface that `remotedialer.connection` satisfies, and those methods delegate to `remotedialer.buffer`. Therefore they could very well be called by higher-level constructs such as http RoundTrippers (eg. Transports) - not necessarily in the `remotedialer` or `rancher/rancher` codebases.

In my local tests, those methods are never called, but I do not have a sure way to tell if they might be in other occasions. In any case, I guess we should strive to implement `net.Conn` in a correct way for future-proofing the code at least.

Reply:
> Should I add that to this patch, or will you create a follow-up PR?

Do you think this could be part of `SURE-7122` since the end result of such race condition would be the goroutine being blocked, so a leak (until something calls `Close()` or `Offer()`? If not, then it should be a separate PR but we'd probably need to create a GH issue for this.

> In any case, I guess we should strive to implement net.Conn in a correct way for future-proofing the code at least.

In that case, should `connection.Read` return a `os.ErrDeadlineExceeded` error if the deadline has exceeded? Probably not needed though for the new timeout introduced here.

Reply:
> Do you think this could be part of SURE-7122 since the end result of such race condition would be the goroutine being blocked, so a leak (until something calls Close() or Offer()? If not, then it should be a separate PR but we'd probably need to create a GH issue for this.

We will soon know - user was given the patch in the original version of PR as a debug image and will report back goroutine profiles before and after the image is installed. In a few days, we should be able to see if a leak still occurs or not:
 - if the leak has stopped, which is my expectation, then it's a separate problem and I will open an issue for it
 - if a leak still occurs, I'll investigate more into the possibility that this particular case is hit, figure out a way to know for sure, and add the relative patch to this PR before it's merged

Is that OK with you?

> > I guess we should strive to implement net.Conn in a correct way for future-proofing the code at least.

> In that case, should connection.Read return a os.ErrDeadlineExceeded error if the deadline has exceeded? Probably not needed though for the new timeout introduced here.

That is correct, I just looked up https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/net/net.go;l=140-141

Fixed in e9c05c8